packages/edit-site/src/components/global-styles/global-styles-provider.js

```diff
@@ -30,6 +33,59 @@ export function mergeBaseAndUserConfigs( base, user ) {
 	} );
 }
 
+// Resolves shared block style variation definitions from the user origin
+// under their respective block types and registers the block style if required.
+function useResolvedBlockStyleVariationsConfig( userConfig ) {
+	const getBlockStyles = useSelect( ( select ) => {
+		return select( blocksStore ).getBlockStyles;
+	}, [] );
+
+	return useMemo( () => {
+		const sharedVariations = userConfig?.styles?.blocks?.variations;
+
+		if ( ! sharedVariations ) {
+			return userConfig;
+		}
+
+		const variationsConfig = cloneDeep( userConfig );
+
+		Object.entries( sharedVariations ).forEach(
+			( [ variationName, variation ] ) => {
+				if ( ! variation?.supportedBlockTypes ) {
+					return;
+				}
+
+				variation.supportedBlockTypes.forEach( ( blockName ) => {
+					// Register any block style variations that have been added
+					// by a theme style variation and are not already registered.
+					const blockStyles = getBlockStyles( blockName );
+					const registeredBlockStyle = blockStyles.find(
+						( { name } ) => name === variationName
+					);
+
+					if ( ! registeredBlockStyle ) {
+						registerBlockStyle( blockName, {
+							name: variationName,
+							label: variationName,
+						} );
+					}
+
+					const path = [
+						'styles',
+						'blocks',
+						blockName,
+						'variations',
+						variationName,
+					];
+					setNestedValue( variationsConfig, path, variation );
+				} );
+			}
+		);
+
+		return deepmerge( variationsConfig, userConfig );
+	}, [ userConfig, getBlockStyles ] );
```

Comment:
> Should this useMemo recalculate when the result of getBlockStyles changes? Or is that not expected to change?

The merged user origin data shouldn't change at all based on what block styles have been registered.  This was a quick fix to the [issue](https://github.com/WordPress/gutenberg/pull/57908#issuecomment-2051034538) Rich flagged and I clearly didn't put enough thought into it.

As noted in my [reply](https://github.com/WordPress/gutenberg/pull/57908#discussion_r1568218825) below, I could probably move this out of this particular `useMemo`.

> I also wonder if it should recalculate when the entire userConfig changes, or only when userConfig?.styles?.blocks?.variations changes.

Good question. It would probably also need to at least have the entirety of `userConfig?.styles?.blocks` as its dependency as any user-selected global styles for a block style variation would need to be merged.

> I'm personally not sure how often it's supposed to trigger

The purpose of this hook is to get the block style variation definitions into their respective block type nodes within the global styles data (and register the block style if it hasn't been already e.g. new variations from a theme style variation).

It would need to trigger when the block style variation definitions changed e.g. when switching theme style variations which include their own new or customized block style variations.

If a user makes changes to a block style variation via global styles, those changes need to be maintained also.

> I'm not sure on the origin of the data.

FWIW the user origin data is retrieved via `useGlobalStylesUserConfig` and the underlying global styles entity record.

My tentative plan here for the moment will be:
1. To move the block style registration to a separate `useEffect`
2. Fix the current known issue around block style variation block type styles not applying properly
3. Then circle back to updating this `useMemo` further so I can more easily spot any breakage I cause

Reply:
> Should this useMemo recalculate when the result of getBlockStyles changes? Or is that not expected to change?

Thinking some more on this, it seems ok as it is, though I think it could be improved by introducing a `hasBlockStyle` selector and using that instead.

Reply:
>  it seems ok as it is

@talldan as it was when you reviewed this, or as it is now, because I'd already pushed a change to this hook a while before this comment?

Reply:
It's ok before and after 👍